internal/allocator/allocator.go

```diff
@@ -345,13 +337,57 @@ func (a *Allocator) sortedPoolsForService(svc *v1.Service) []*config.Pool {
 	return append(priorityPools, noPriorityPools...)
 }
 
+func (a *Allocator) sortedNamespacePoolsForService(svc *v1.Service) ([]*config.Pool, []*config.Pool) {
+	var priorityPools, noPriorityPools []*config.Pool
+	svcLabels := labels.Set(svc.Labels)
+	for _, nsPoolName := range a.pools.ByNamespace[svc.Namespace] {
+		if nsPool, ok := a.pools.ByName[nsPoolName]; ok {
+			if !nsPool.AutoAssign {
+				continue
+			}
+			hasSvcLabelMatch := false
+			for _, svcSelector := range nsPool.ServiceAllocations.ServiceSelectors {
+				if svcSelector.Matches(svcLabels) {
+					hasSvcLabelMatch = true
+					break
+				}
+			}
+			if len(nsPool.ServiceAllocations.ServiceSelectors) > 0 && !hasSvcLabelMatch {
+				continue
+			}
+			if nsPool.ServiceAllocations.Priority > 0 {
+				priorityPools = append(priorityPools, nsPool)
+				continue
+			}
+			noPriorityPools = append(noPriorityPools, nsPool)
+		}
+	}
+	sort.Slice(priorityPools, func(i, j int) bool {
+		return priorityPools[i].ServiceAllocations.Priority <
+			priorityPools[j].ServiceAllocations.Priority
+	})
+	return priorityPools, noPriorityPools
+}
+
 func (a *Allocator) isPoolCompatibleWithService(p *config.Pool, svc *v1.Service) bool {
-	if p.ServiceAllocations != nil && p.ServiceAllocations.Namespaces.Len() > 0 &&
-		!p.ServiceAllocations.Namespaces.Has(svc.Namespace) {
-		return false
+	svcLabels := labels.Set(svc.Labels)
+	if p.ServiceAllocations != nil && p.ServiceAllocations.Namespaces.Len() > 0 {
+		if !p.ServiceAllocations.Namespaces.Has(svc.Namespace) {
+			return false
+		}
+		hasSvcLabelMatch := false
```

Comment:
Is this necessary? If we don't return false here we'll fall back in the check below, and if the service doesn't match we'll return false anyway

Reply:
yes, this check is needed for service allocations which has only namespaces, but not service selectors.

Reply:
but if a service has no service selectors we'll go to the final return, right?